osh/builtin_assign.py

```diff
@@ -25,10 +26,122 @@
   from frontend import arg_def
 
 
+def _PrintVariables(mem, cmd_val, arg, print_flags, readonly=False, exported=False):
+  # type: (Mem, value_t, Any, bool, bool, bool) -> int
+  flag_g = getattr(arg, 'g', None)
+  flag_n = getattr(arg, 'n', None)
+  flag_r = getattr(arg, 'r', None)
+  flag_x = getattr(arg, 'x', None)
+  flag_a = getattr(arg, 'a', None)
+  flag_A = getattr(arg, 'A', None)
+
+  lookup_mode = scope_e.Dynamic
+  if cmd_val.builtin_id == builtin_i.local:
+    if flag_g and not mem.IsGlobalScope():
+      return 1
+    lookup_mode = scope_e.LocalOnly
+  elif flag_g:
+    lookup_mode = scope_e.GlobalOnly
+
+  if len(cmd_val.pairs) == 0:
+    print_all = True
+    cells = mem.GetAllCells(lookup_mode)
+    names = sorted(cells)
+  else:
+    print_all = False
+    names = []
+    cells = {}
+    for pair in cmd_val.pairs:
+      name = pair.lval.name
+      if pair.rval and pair.rval.tag_() == value_e.Str:
+        name += "=" + cast(value__Str, pair.rval).s
+        names.append(name)
+        cells[name] = None
+      else:
+        names.append(name)
+        cells[name] = mem.GetCell(name, lookup_mode)
+
+  count = 0
+  for name in names:
+    cell = cells[name]
+    if cell is None: continue
+    val = cell.val
+
+    if val.tag_() == value_e.Undef: continue
+    if readonly and not cell.readonly: continue
+    if exported and not cell.exported: continue
+    if flag_n == '-' and not cell.nameref: continue
+    if flag_n == '+' and cell.nameref: continue
+    if flag_r == '-' and not cell.readonly: continue
+    if flag_r == '+' and cell.readonly: continue
+    if flag_x == '-' and not cell.exported: continue
+    if flag_x == '+' and cell.exported: continue
+    if flag_a and val.tag_() != value_e.MaybeStrArray: continue
+    if flag_A and val.tag_() != value_e.AssocArray: continue
+
+    decl = []
+    if print_flags:
+      flags = []
+      if cell.nameref: flags.append('n')
+      if cell.readonly: flags.append('r')
+      if cell.exported: flags.append('x')
+      if val.tag_() == value_e.MaybeStrArray:
+        flags.append('a')
+      elif val.tag_() == value_e.AssocArray:
+        flags.append('A')
+      if len(flags) == 0: flags.append('-')
+
+      decl.extend(["declare -", ''.join(flags), " ", name])
```

Comment:
It looks like you don't need `''.join(flags)` or `''.join(body)` below, because we join it at the end anyway.  But this is a minor detail

Reply:
Just for curiosity, how can I write in this case? The reason why I maintained other lists `flags` and `body` is because I wanted to use `len(flags)` and `len(body)`. Ah, OK. Maybe I could write something like the following, but I just didn't want to split a simple line into many lines.

```py
decl.append("declare -")
decl.extend(flags)
decl.extend([" ", name])
```

Or this? But the following code seems even less efficient because it induces addtional list generations.
```py
decl.extend(["declare -"] + flags + [" ", name])
```

Reply:
I usually check guard space separation in a loop with `i != 0` , and then you don't need to check len(body).  And I think you can check `len(assoc_val.d)` for the last check.

But those are minor details and I think it's fine.  There is one more string allocation for `''.join(flags)` and one more for `''.join(body)`, but like you say it's easy to introduce other allocations elsewhere.  After translating to C++ the previous style of `+=` would introduce a lot more allocations.

I actually did go back and optimize allocations in C++ generated from Python for the parser!  e.g. the last step of http://www.oilshell.org/blog/2020/01/parser-benchmarks.html

But I think this is pretty close to optimal, and it's not a hot path anyway, i.e. `declare -p` is not that common.

Reply:
OK! I see, it is aimed to be translated to C++. Thank you!